semaphore.ts

```diff
@@ -31,46 +41,53 @@ export class Semaphore {
         `size must be a positive safe integer, got ${size}`,
       );
     }
-    this.#rest = size + 1;
+    this.#value = size;
   }
 
   /**
    * Returns true if the semaphore is currently locked.
    */
   get locked(): boolean {
-    return this.#rest === 0;
+    return this.#value === 0;
   }
 
   /**
-   * Acquires a lock on the semaphore, and invokes the specified function.
+   * Acquires a lock and invokes the specified function.
    *
    * @param fn The function to invoke.
    * @returns A promise that resolves to the return value of the specified function.
    */
   async lock<R>(fn: () => R | PromiseLike<R>): Promise<R> {
-    await this.#acquire();
-    try {
-      return await fn();
-    } finally {
-      this.#release();
-    }
+    using _lock = await this.acquire();
+    return await fn();
   }
 
-  async #acquire(): Promise<void> {
-    if (this.#rest > 0) {
-      this.#rest -= 1;
-    }
-    if (this.#rest === 0) {
-      await this.#notify.notified();
+  /**
+   * Acquire a lock and return a promise with disposable that releases the a lock when disposed.
+   *
+   * @returns A Promise with Disposable that releases the mutex when disposed.
+   */
+  acquire(): Promise<Disposable> & Disposable {
```

Comment:
so this seems to enable an interesting error case — it is possible to release the same lock twice.

```typescript
using _pending = sem.acquire(); // will dispose the disposable promise
// some stuff
using _lock = await pending; // the inner lock will *also* be disposed
```

Now, that seems unusual and I don't think people should do it, but making both the promise value and the promise itself `Disposable` allows for it. ISTM that the intent here is to allow a pending lock attempt to be cancelled, but double-release protection of some kind is needed (or "cancel" and "release" should be different actions).

Reply:
It is not clear why such a design and order is needed at all

What's the problem with using it like this?
Yes, and this does not solve the problem of unnecessary calls using

If you need protection against the repeated return of a single lock, then you only need to set the Disposable state to the object

```ts
import { scheduler } from 'node:timers/promises';
import { Semaphore } from "./semaphore.ts";

const sem = new Semaphore(1);

async function doSome(obj: {text: string}) {
	const pendingLock = sem.acquire();

	// do task ...
	await scheduler.wait(1000);

	using _lock = await pendingLock;
	//maybe throw 
	obj.text += 'hello '

	using _lock2 = await pendingLock;
	//maybe throw 
	obj.text += 'world'

	return obj;
}
```

Reply:
And I misunderstood the question